### PR TITLE
Implement SIP-42 - Support for binary integer literals

### PR DESCRIPTION
**Pull request by Sporarum**
Implement, test and document SIP-42

(Implement is a strong word, I uncommented a single line)

### CODE REVIEW

File: tests/run/binaryLiterals.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
+@main
+def Test =
+  assert(0b0000 == 0)
```
Comment (bishabosha) [inline tests/run/binaryLiterals.scala:3]: didn't test `0B...` variant

Comment (bishabosha) [reply tests/run/binaryLiterals.scala:9]: perfect :)

File: tests/run/binaryLiterals.scala [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  assert(0b0001_0000_0000_0000__0000_0000_0000_0000 == 268435456)
+  assert(0b0010_0000_0000_0000__0000_0000_0000_0000 == 536870912)
+  assert(0b0100_0000_0000_0000__0000_0000_0000_0000 == 1073741824)
+  assert(0b1000_0000_0000_0000__0000_0000_0000_0000L == 2147483648L)
+
+  assert(0b1000_0000_0000_0000__0000_0000_0000_0000 == -2147483648) // Signed !
+  assert(0b1111_1111_1111_1111__1111_1111_1111_1111 == -1)
+
+  // Randomly generated using https://numbergenerator.org/random-32-bit-binary-number#!numbers=10&length=32&addfilters=
+  // Converted to sined decimal using https://onlinetoolz.net/unsigned-signed#base=2&bits=32
```
Comment (ritschwumm) [inline tests/run/binaryLiterals.scala:54]: typo `sined`

Comment (Sporarum) [reply tests/run/binaryLiterals.scala:54]:
```suggestion
  // Converted to signed decimal using https://onlinetoolz.net/unsigned-signed#base=2&bits=32
```

**Comment by odersky** [comment]
Is it in the Scala 2 spec? If not we need a doc page for the change. As often docs are harder than code.

**Comment by som-snytt** [comment]
I had forgotten that I forward-ported the commented code with other fixes.

The old NthPortal PR was https://github.com/scala/scala/pull/8275/files which had only the syntax summary tweak. It has some more tests.

Nth's comment:
```
what I actually should have tested is 0b1 // Kenobi
```
That effort was in the six months leading up to the pandemic. My opinion is that it was unfortunate to put Scala 3 uptake in the path of Scala 2. That is, `0b` syntax could have benignly coexisted. It would have blocked no one's migration and confused no beginners. And we would have had four years to come up with a literal syntax pun on `n00b`.

**Comment by bishabosha** [comment]
here is the relevant doc section to update: https://scala-lang.org/files/archive/spec/3.4/01-lexical-syntax.html#literals

**Comment by odersky** [comment]
Plus, we should have a short doc page in the docs/_docs/reference pages explaining the feature with a short example. We do that for all features that are new in Scala 3.

**Comment by Sporarum** [comment]
I was expecting it to get backported to Scala 2, since there is already an implementation targetting it, will this not be the case ?

**Comment by som-snytt** [comment]
@Sporarum that is likely. https://github.com/scala/scala/pull/10656

**Comment by odersky** [comment]
Bit in any case we need a doc page to explain the new feature, whether it will be in Scala 2 or not. It should not be hard...

**Comment by Sporarum** [comment]
@odersky
Totally agree, and I'll try to do it this weekend, it's more a question of where, and thus in which format ?

**Comment by odersky** [comment]
I think it should have an entry here:

https://dotty.epfl.ch/docs/reference/index.html

Local directory is docs/_docs/reference.

**Comment by Sporarum** [comment]
I have added a short reference page, let me know if something is missing

File: docs/_docs/reference/changed-features/binary-literals.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+---
+layout: doc-page
+title: "Binary Integer Literals"
+nightlyOf: https://docs.scala-lang.org/scala3/reference/changed-features/binary-integer-literals.html
+---
+
```
Comment (bishabosha) [inline docs/_docs/reference/changed-features/binary-literals.md:6]: I think you also need to update the sidebar manually

Comment (Sporarum) [reply docs/_docs/reference/changed-features/binary-literals.md]: Oh right !

File: docs/sidebar.yml [hunk 4]
```diff
@@ -81,6 +81,7 @@ subsection:
           - page: reference/other-new-features/safe-initialization.md
           - page: reference/other-new-features/type-test.md
           - page: reference/other-new-features/experimental-defs.md
+          - page: reference/other-new-features/binary-literals.md
```
Comment (bishabosha) [inline docs/sidebar.yml:84]: and there's another sidebar for the "stable reference" 😅

Comment (som-snytt) [reply docs/sidebar.yml:84]: Surprised to see the PR still open, as it's just 0's and 1's, but it's the dread "long tail of documentation".

**Review by sjrd — APPROVED** [review]

File: docs/sidebar.yml (hunk 4 above)
Comment (Sporarum) [reply docs/sidebar.yml:84]: Okay, I'll fix that as well !

**Comment by Sporarum** [comment]
I added the line to the second sidebar, and updated the Scala 3 Spec (I mistook the earlier message with one speaking about the Scala 2 Spec)

If it's good like that, let me know and I'll squash my commits together

**Comment by Sporarum** [comment]
@bishabosha, @sjrd I wasn't sure if I could merge it, so I'll leave you to it

**Comment by nicolasstucki** [comment]
I noticed we are missing a test for `0b1010l`.

**Pull request merged**